### PR TITLE
refactor(snapshot): use draw buffer interface

### LINKED ISSUE #5261: Use draw_buf in API  lv_snapshot_take_to_buf(obj, draw_buf)
Let's change the API `lv_snapshot_take_to_buf ` to lv_snapshot_take_to_buf(obj, draw_buf) in future.
For now you can use `lv_draw_buf_from_image(&draw_buf, dsc);` directly.

Unified draw buffer is much easier to use instead of raw pointers.

_Originally posted by @XuNeo in https://github.com/lvgl/lvgl/pull/5254#discussion_r1447094941_

### PR DESCRIPTION
**Pull request by XuNeo**
### Description of the feature or fix

`lv_snapshot_take_to_buf` should use lv_draw_buf_t instead of assembling the struct manually.

Fix #5261

The issue exposed in https://github.com/lvgl/lvgl/pull/5483 suggests that we should avoid converting between `lv_image_dsc_t` and `lv_draw_buf_t`.
The rule is that `lv_image_dsc_t` is for constant C-array image assets. And `lv_draw_buf_t` is for image used in run-time.

`_lv_snapshot_take_to_buf` and `lv_snapshot_free` for legacy use.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html)
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: examples/others/snapshot/lv_example_snapshot_1.c [hunk 1]
```diff
@@ -7,7 +7,7 @@ static void event_cb(lv_event_t * e)
     lv_obj_t * img = lv_event_get_target(e);
 
     if(snapshot_obj) {
-        lv_image_dsc_t * snapshot = (void *)lv_image_get_src(snapshot_obj);
+        lv_draw_buf_t * snapshot = (void *)lv_image_get_src(snapshot_obj);
```
Comment (W-Mai) [inline examples/others/snapshot/lv_example_snapshot_1.c]:
```suggestion
        lv_draw_buf_t * snapshot = (lv_draw_buf_t *)lv_image_get_src(snapshot_obj);
```

**Review by W-Mai — CHANGES_REQUESTED** [review]

File: src/others/snapshot/lv_snapshot.h [hunk 2]
```diff
@@ -63,8 +93,22 @@ uint32_t lv_snapshot_buf_size_needed(lv_obj_t * obj, lv_color_format_t cf);
  * @param buf_size  provided buffer size in bytes.
  * @return          LV_RESULT_OK on success, LV_RESULT_INVALID on error.
  */
-lv_result_t lv_snapshot_take_to_buf(lv_obj_t * obj, lv_color_format_t cf, lv_image_dsc_t * dsc, void * buf,
-                                    uint32_t buf_size);
+static inline lv_result_t _lv_snapshot_take_to_buf(lv_obj_t * obj, lv_color_format_t cf, lv_image_dsc_t * dsc,
```
Comment (XuNeo) [inline src/others/snapshot/lv_snapshot.h]: Should we remove the legacy API directly? This usage should be forbidden.

File: src/others/snapshot/lv_snapshot.h [hunk 3]
```diff
@@ -37,24 +37,54 @@ extern "C" {
...
-lv_image_dsc_t * lv_snapshot_take(lv_obj_t * obj, lv_color_format_t cf);
+lv_draw_buf_t * lv_snapshot_take(lv_obj_t * obj, lv_color_format_t cf);
 
 /**
  * Free the snapshot image returned by @ref lv_snapshot_take
  * It will firstly free the data image takes, then the image descriptor.
  * @param dsc   the image descriptor generated by lv_snapshot_take.
  */
-void lv_snapshot_free(lv_image_dsc_t * dsc);
+void lv_draw_buf_destroy(lv_draw_buf_t * dsc);
```
Comment (FASTSHIFT) [inline src/others/snapshot/lv_snapshot.h]: remove

File: src/others/snapshot/lv_snapshot.h [hunk 4]
```diff
@@ -37,24 +37,54 @@ extern "C" {
...
 
 /**
- * Get the buffer needed for object snapshot image.
- * @param obj   the object to generate snapshot.
- * @param cf    color format for generated image.
- * @return      the buffer size needed in bytes
+ * Create a draw buffer for object to store the snapshot image.
+ */
```
Comment (kisvegabor) [inline src/others/snapshot/lv_snapshot.h:47]: Add comment for the parameters.

File: src/others/snapshot/lv_snapshot.h [hunk 5]
```diff
@@ -37,24 +37,54 @@ extern "C" {
...
+lv_draw_buf_t * lv_snapshot_draw_buf_create(lv_obj_t * obj, lv_color_format_t cf);
+
+/**
+ * Reshape the draw buffer to prepare for take snapshot for obj.
+ * This is usually used to check if the existing draw buffer is enough for
+ * obj snapshot. If return LV_RESULT_INVALID, you should create a new one.
+ *
```
Comment (kisvegabor) [inline src/others/snapshot/lv_snapshot.h]:
```suggestion
 
```

File: src/others/snapshot/lv_snapshot.h [hunk 6]
```diff
@@ -37,24 +37,54 @@ extern "C" {
...
+void lv_draw_buf_destroy(lv_draw_buf_t * dsc);
 
 /**
- * Get the buffer needed for object snapshot image.
- * @param obj   the object to generate snapshot.
- * @param cf    color format for generated image.
- * @return      the buffer size needed in bytes
+ * Create a draw buffer for object to store the snapshot image.
+ */
+lv_draw_buf_t * lv_snapshot_draw_buf_create(lv_obj_t * obj, lv_color_format_t cf);
```
Comment (kisvegabor) [inline src/others/snapshot/lv_snapshot.h]:
```suggestion
lv_draw_buf_t * lv_snapshot_create_draw_buf(lv_obj_t * obj, lv_color_format_t cf);
```

File: src/others/snapshot/lv_snapshot.h [hunk 7]
```diff
@@ -37,24 +37,54 @@ extern "C" {
...
+
+/**
+ * Reshape the draw buffer to prepare for take snapshot for obj.
+ * This is usually used to check if the existing draw buffer is enough for
+ * obj snapshot. If return LV_RESULT_INVALID, you should create a new one.
+ *
+ * @param draw_buf  the draw buffer to reshape.
+ * @param obj       the object to generate snapshot.
+ */
+lv_result_t lv_snapshot_draw_buf_reshape(lv_obj_t * obj, lv_draw_buf_t * draw_buf);
```
Comment (kisvegabor) [inline src/others/snapshot/lv_snapshot.h]:
```suggestion
lv_result_t lv_snapshot_reshape_draw_buf(lv_obj_t * obj, lv_draw_buf_t * draw_buf);
```

File: src/others/snapshot/lv_snapshot.h [hunk 8]
```diff
@@ -37,24 +37,54 @@ extern "C" {
...
+
+/**
+ * Take snapshot for object with its children, save image info to provided buffer.
+ * @param obj       the object to generate snapshot.
+ * @param cf        color format for new snapshot image.
+ *                  It could differ with cf of `draw_buf` as long as the new cf will fit in.
+ * @param draw_buf  the draw buffer to store the image result. It's reshaped automatically.
+ * @return          LV_RESULT_OK on success, LV_RESULT_INVALID on error.
+ */
+lv_result_t lv_snapshot_take_to_buf(lv_obj_t * obj, lv_color_format_t cf, lv_draw_buf_t * draw_buf);
```
Comment (kisvegabor) [inline src/others/snapshot/lv_snapshot.h]:
```suggestion
lv_result_t lv_snapshot_take_to_draw_buf(lv_obj_t * obj, lv_color_format_t cf, lv_draw_buf_t * draw_buf);
```

File: src/others/snapshot/lv_snapshot.h (hunk 2 above)
Comment (kisvegabor) [reply src/others/snapshot/lv_snapshot.h]: I think it's enough provide an example in the docs.

Comment (XuNeo) [reply src/others/snapshot/lv_snapshot.h]: Since the new API name is different, let keep the legacy API here.

**Review by kisvegabor — APPROVED** [review]

**Comment by kisvegabor** [comment]
Thank you! 👍

**Pull request merged**